### DISCUSSION TITLE
How to sort by one axis in a dual-y axis bar chart?

### DISCUSSION THREAD

**Discussion by woter1832** [vega/vega#3993] [Q&A]
Hi,

I've created the following visual in Power BI using Vega via the Deneb plugin.

The sort is not working as I would expect. I want the cost (blue) bars to go from small to large (left to right), but as you can see this isn't happening.

![image](https://github.com/user-attachments/assets/aa0378f4-81cb-4275-9967-e8810a7a8f0e)

Here is the code:
```
{
  "data": [
    {
      "name": "dataset",
      "transform": [
        {
          "type": "collect",
          "sort":{"field":"Cost", "order":"ascending"}
        }
        
      ]
    }
  ],
  
  "scales": [
    {
      "name": "xscale",
      "type": "band",
      "domain": {"data": "dataset", "field": "Model"},
      "range": "width",
      "padding": 0.4,
      "round": true
    },
    {
      "name": "yscale1",
      "type": "linear",
      "domain": {"data": "dataset", "field": "Score"},
      "range": "height",
      "nice": true
    },
    {
      "name": "yscale2",
      "type": "linear",
      "domain": {"data": "dataset", "field": "Cost"},
      "range": "height",
      "nice": true
    }
  ],
  "axes": [
    {"orient": "bottom", "scale": "xscale", "title": "Model"},
    {"orient": "left", "scale": "yscale1", "title": "Score"},
    {"orient": "right", "scale": "yscale2", "title": "Cost", "format": "$.0f"}
  ],
  "marks": [
    {
      "type": "rect",
      "from": {"data": "dataset"},
      "encode": {
        "enter": {
          "x": {"scale": "xscale", "field": "Model"},
          "width": {"scale": "xscale", "band": 1},
          "y": {"scale": "yscale1", "field": "Score"},
          "y2": {"scale": "yscale1", "value": 0},
          "tooltip": {"field": "Cost"}
          
          
        },
        "update": {
          "fill": {"value": "steelblue"},
          "href": {"signal": "datum.link"}

        },
        "hover": {"fill": {"value": "blue"}}
      }
    },
    {
      "type": "rect",
      "from": {"data": "dataset"},
      "encode": {
        "enter": {
          "x": {"scale": "xscale", "field": "Model", "offset": 21},
          "width": {"scale": "xscale", "band": 0.5},
          "y": {"scale": "yscale2", "field": "Cost"},
          "y2": {"scale": "yscale2", "value": 0},
          "tooltip": {"field": "Score"}
          
        },
        "update": {
          "fill": {"value": "orange"},
          "href": {"signal": "datum.link"}

        },
        "hover": {"fill": {"value": "darkorange"}}
      }
    }
  ]
}
```

I found [this](https://stackoverflow.com/questions/79016846/power-bi-deneb-vega-lite-sorting-specification), but I couldn't get either suggestion to work, and then I realised it's Vega-lite. I tried to produce the same graph in Vega-lite, but couldn't get it to work :cry:

Any help would be greatly appreciated.

Thanks

**Comment by hydrosquall** [comment]
Hi @woter1832 - it will be difficult for us to debug this graph without a spec that includes sample data. Could you share a link to your graph with values filled in using the vega editor https://vega.github.io/editor/#/custom/vega-lite ? Alternately, providing the CSV or JSON with the data can work too.

I'm optimistic that the graph you want to create  will be doable more simply with Vega-Lite rather than Vega, but it will help to have a sample dataset to reference first.

**Comment by woter1832** [comment]
Hi @hydrosquall 

I gave up on PowerBI and tried it natively, although I'd guess if it worked natively, it would work in PowerBI.

Apologies for the delay. In the end, I used [PowerPoint](https://thingsgeeky.arcotek.co.uk/public/computing/devices/euc/laptop-selection.html#chart) (yuk 😉 ). However, that doesn't mean I don't want to learn how it can be done.

I gave up on PowerBI and tried it natively, although I'd guess if it worked natively, it would work in PowerBI.

I made several attempts using both Vega and Vega-Lite.

- The data is in a Gist [here](https://gist.github.com/woter1832/aa1fb67047ce77e1d48a698cd0828c10).
- A selection of Vega-Lite Gists [here](https://gist.github.com/woter1832/), where I'm trying different things.


I'd be very grateful if you could show me how to achieve this as I can see Vega is a powerful tool.

Thanks in advance.

W.

**Comment by hydrosquall** [comment]
Hi @woter1832 ,

<img width="820" alt="image" src="https://github.com/user-attachments/assets/5f770304-bd8d-4be6-982f-6226ef64ab18" />

I think I've recreated your graph using Vega Lite, sorting by the blue (Cost) column descending.

There is one exception: since your CSV has repeats for the `Model` field, (I.E the HP Zbook Fury G11), those bars are stacked along the Y axis using a sum. I assume that since you have control of the data on your side, you can either deduplicate them, make the graph based on "product ID" rather than "model", or change the aggregator to avg/max/min, whatever you need.

Here's an [editor link](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUAwBOgtBmx5CBWhEzU6jJgwhxtUOZjizzDhAQDubJ9oCMADgBmACYCJCRfADMmADYAdgAGABY4qDg4uLhfZST-JBiATn8oZQT-YOLfBIJtJA8CSLE0pJjlAJjImKgg2OCW8qTI31SypiYxJEicguCqgkEkHEw2HH5bQTgoTHZZfiYkCBooSigIYhAAXyVnB2UaWTI0UAAPJ5BImjhBZX0AWTZlF9FCBMABPHBwfSyNgIe5IPRKGTaTBvD5fH7oADCbDMwLY2kB2n0gOgzjuD0uSiQzxoEDeCyYXwAgg8NmgEhdOUoFqC7GgANqgZDaADWbzBEP0+yJVxANwB90emhAoNRn2++mxuKUEsh6AAjgwkC46GoaKRgdTaaj8cgUegwJQEpE8dpPi59G6yFhLpyrkKkKLxeC9SBpcCHIJ8fp8cayJClB4aMp6GhAhzrrJboq3qrlWiNegAMoObQJkEh-SG43bdTbC1Uml05X4932kAbSIov0XAC6SjLMkEFuV0HhetAeZA90BENkgI9XJASZTMDQ-gSCSU8Bo3vt6Y5QA)

With respect to your original efforts, I think you can avoid the collect transform

```diff
        {
-          "type": "collect",
-          "sort":{"field":"Cost", "order":"ascending"}
        }
```

Instead, you can put the sort directly on the x scale. Here's how it looks in Vega: 

```diff
  "scales": [
    {
      "name": "x",
      "type": "band",
      "domain": {
        "data": "source_0",
        "field": "Model",
+        "sort": {"op": "sum", "field": "Cost", "order": "descending"}
      },
      "range": [0, {"signal": "width"}],
      "paddingInner": 0.1,
      "paddingOuter": 0.05
    },
   ... etc
```

Personally, I would stick with Vega-Lite for this graph, I don't think you need to be writing full Vega.

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "data": {
    "url": "https://gist.githubusercontent.com/woter1832/aa1fb67047ce77e1d48a698cd0828c10/raw/f5ce46d186f6c83b6246824f17c08bb5afd49210/laptop-selection-basic.csv"
  },
  "encoding": {
    "x": {
      "field": "Model",
      "type": "nominal",
      "sort": {
        "field": "Cost",
        "order": "descending"
      },
      "axis": {
        "labelAngle": 0
      }
    }
  },
  "layer": [
    {
      "mark": {
        "type": "bar"
      },
      "encoding": {
        "y": {
          "field": "Cost",
          "type": "quantitative",
          "axis": {
            "format": "$.0f",
            "orient": "right"
          }
        }
      }
    },
    {
      "mark": {
        "type": "bar",
        "color": "orange",
        "width": 30
      },
      "encoding": {
        "y": {
          "field": "Score",
          "type": "quantitative",
          "axis": {
            "orient": "left"
          }
        }
      }
    }
  ],
  "resolve": {
    "scale": {
      "y": "independent"
    }
  },
  "width": 800,
  "height": 300
}
```

Let me know if this works for you.

**Reply by woter1832** [reply]
Thank you @hydrosquall, looking good. Now you mention it, a penny has dropped regarding duplicate models. As you say, I should use something unique like product code.

One of the issues I was having was with the scale. The most expensive laptop is ~£5000, where the graph shows it as ~£13000. I'm on my phone, so it's difficult to confirm, but from memory, the highest score was about 300.

Something else that I couldn't get to work was using the £ (GBP) sign as opposed to $ sign. I was reading that it should work it out from the locale.

Thanks again. I'll have a detailed look on Monday.

W.

**Comment by hydrosquall** [comment]
Glad the example was helpful!

I think the 13000 price appears because the models that appear multiple
times are having their values summed.

We can use xOffset channel if you wish to make horizontally grouped bar
chart to allow for multiple product ids per model, but it might change your
sort. The simplest fix would be to remove the duplicates

About currency -

- if you are in PowerBI, use pbiFormat :
https://kerrykolosko.com/formatting-values-and-tooltips-in-deneb/
- if you are in a custom page (using Vega-embed), see
https://groups.google.com/g/vega-js/c/2j2oXkvJ6JI
